src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSigner.cs

```diff
@@ -313,19 +378,60 @@ internal SignerInfoAsn Sign(
 
             if (SignerIdentifierType == SubjectIdentifierType.NoSignature)
             {
+                // The behavior of this scenario should match Windows which currently does not
+                // implement PQC. So we do a best effort determination of whether the algorithm
+                // is a pure algorithm and throw if so. This is subject to change once Windows
+                // implements PQC.
+                string? keyAlgorithm = null;
+                if (Certificate != null)
+                {
+                    try
+                    {
+                        keyAlgorithm = Certificate.GetKeyAlgorithm();
+                    }
+                    catch (CryptographicException)
+                    {
+                    }
+                }
+
+                if (keyAlgorithm != null)
+                {
+                    CmsSignature? processor = CmsSignature.ResolveAndVerifyKeyType(keyAlgorithm, _privateKey, SignaturePadding);
+                    if (processor?.NeedsHashedMessage == false)
```

Comment:
I can definitely imagine that this was an undertested area before...but does NoSignature currently work for unknown algorithms?  (Do we have tests for that?)

I'm just surprised that this path is allowing ResolveAndVerifyKeyType to return null.